Support for other conversion currencies and Improved Guide

paras-lehana opened the pull request:
## Contribution (Paras Lehana)

> Fork by paras-lehana

**Tweaks:**

* I have changed the decimal precison to none. If you want to include, remove `int` conversion in `self.widget.set_label` in `app/widgets.py` and change precison from 0 to (for example, back to) 2.
* You can add other configuration in example.yaml file by adding more commas. Be sure to use the fields in `__init__` in `app/widgets.py` after conf has been split. 
* Renamed `tokens-example.yaml` to `tokens.yaml` by default. Removed `tokens.yaml` from `.gitignore` as well due to added inline guide. Consider adding it again if you are pushing changes not related to it. 
* Do contact me for any more suggestion or additions. I'll be happy to add those. I'm grateful to peter-featherstone for this awesome widget.  

**Improved Guide for defining Tokens in tokens.yaml:**

* You can find brief guide to help yourself adding new tokens in tokens.yaml.
* You need to define Token ID in the file as CMC API doesn't support symbols yet. 
* Token ids are generally same as coin name in lowercase. Replace space by hyphen in multiword token names. For example, `basic-attention-token` for Basic Attention Token (BAT)
* Find exact ids on CMC's token page url. For example, `coinmarketcap.com/currencies/basic-attention-token/`
* You can change refresh time in seconds by changing `REFRESH_TIME_IN_SECONDS` variable in `app/widget.py` file. Default is 10 mins so you may want to consider changing it. 


**Support for other conversion currencies including Satoshis:**

* I have added support for defining crypto-currencies and fiat as conversion currencies.
* Satoshi is also supported. Use 'SAT' for the same. For example, `bgogo-token,SAT`. 1 BTC equals 100 million Satoshis.
* You can define conversion currency as `<coin-id>,<conversion currency>`. For example, `Ethereum,INR` will show Ethereum price in Indian Rupees. 
* Both crypto-currency and fiat are supported for conversion as per CMC data. For example, you can use `Ripple,EUR` as well as `Ripple,BTC`.
* I have added a local array `currency_symbol` in `app/widget.py` file that countains symbol for different currencies (including crypto). Consider contributing into it or suggest me a suitable python library. I have used `シ` symbol for Satoshi. You change it to whatever you like.

peter-featherstone commented on app/widget.py:
```diff
@@ -9,18 +9,22 @@
-IMAGE_URL = 'https://s2.coinmarketcap.com/static/img/coins/32x32/\d+.png'
-REFRESH_TIME_IN_SECONDS = 600
+IMAGE_URL = 'https://s2.coinmarketcap.com/static/img/coins/64x64/\d+.png'
+REFRESH_TIME_IN_SECONDS = 5
```
According to CoinMarketCaps API docs the free tier is rate limited to 10 calls per minute. This would call the API 20 times per minute. Did you see any issues with this when running it? https://coinmarketcap.com/api/documentation/v1/#section/Errors-and-Rate-Limits

paras-lehana replied:
> According to CoinMarketCaps API docs the free tier is rate limited to 10 calls per minute.

Oh, I missed this. Let's change this to 8 or 10 for being on safer side. I'll do it tomorrow and you can then review the pull request again.

> Did you see any issues with this when running it? 

Actually, on the frontend, obviously, No. I was more into the code. I think 10 second refresh is not an issue as well. I'll run the script and watch out for console outputs/errors - should send a non-200 signal, no?

peter-featherstone replied:
According to the docs it should return a status code of `429` if we hit the limits. Apart from that the changes in this PR look great, thanks for taking the time to make them!

Feel free to make the API update if you'd like too. To be honest I'm surprised to learn people are using this repository as I never thought anyone else was using it apart from me so this is cool!

paras-lehana replied:
Actually many people would need this type of Ticker and upon searching your repo is among the top ones. I have tried many other projects as well but found none of them comparable to yours in terms of feature and code readability. 

The installation may be troublesome for some users and I'm ready to help them with regular documentation.

peter-featherstone replied:
That's really good to hear, I'm glad you like it and have been using it 😄

paras-lehana replied:
@peter-featherstone I have done the required changes in refresh time. Please review.

peter-featherstone replied:
Looks good. Are you happy for me to merge it in now?

paras-lehana replied:
Yes, if you're okay with it. Do let me know for any other suggestions. :)

peter-featherstone replied:
The only suggestion I have is around PEP-8 as I try to follow that as closely I can in my Python work. I know you said you are new to Python so I'm not sure if you have had the chance to read through PEP-8 guidance https://www.python.org/dev/peps/pep-0008/. I'll leave some comments on specific lines regarding some of the points on the PR for you.

tmlee commented:
May i suggest considering our API at https://www.coingecko.com/en/api
It is widely used and battle tested by many production applications out there.

We have flexible rate limit and no keys required to access. If you are looking to display token price, we have support for query using contract/token addresses using our simple API
For example https://api.coingecko.com/api/v3/simple/token_price/ethereum?contract_addresses=0x0d8775f648430679a709e98d2b0cb6250d2887ef%2C0xB8c77482e45F1F44dE1745F52C74426C631bDD52&vs_currencies=usd%2Ceur%2Cgbp

Happy to answer any questions you may have!

peter-featherstone commented on app/widget.py:
```diff
@@ -38,22 +42,21 @@ def _set_label_image(self):
 
         self.widget.set_label(
-            ' ${}'.format(round(float(token_data['price_usd']), 2)),
+            ' {}{}'.format(currency_symbol[self.convert] if self.convert in currency_symbol else '',int(round(float(token_data['price_'+str(self.convert) if self.convert != 'sat' else 'price_btc'])*(1 if self.convert != 'sat' else 100000000),0))),
```
Can you follow the PEP-8 guidance on maximum line length for this part https://www.python.org/dev/peps/pep-0008/#maximum-line-length. It's quite a long line as it currently is.

paras-lehana replied:
Yes, I know I may not have followed python  coding conventions as I was new to the language.

Actually, this line contains the argument to the set_label function so I didn't want to mess with OOP concepts and write the conditional statements (that depends on entries in tokens.yaml file) outside the function. I don't know the exact flow of the code so can you **suggest me** how to break my additions and put it somewhere more suitable. 

I'm pretty sure that you've understood this simple addition but if you have any query about the commit, do tell me and I'm ready to brief you.

peter-featherstone replied:
The best way would be to break it down into smaller, more meaniingful variables. It also makes the code cleaner and the intent more obvious. For example, something like:

```
token_key = 'price_' + str(self.convert) if self.convert != 'sat' else 'price_btc'
amount = token_data[token_key] * (1 if self.convert != 'sat' else 100000000)
' {}{}'.format(
    currency_symbol.get(self.convert, ''), 
    int(round(float(amount), 0))
)
```

Python also has the very useful get method for dicts that you can provide a default value for if the key doesn't exist so `currency_symbol[self.convert] if self.convert in currency_symbol else ''` just becomes `currency_symbol.get(self.convert, '')`

paras-lehana replied:
Actually I had initially written the code in similar way. Since the code was outside the intended function, I thought to ask you first in case I'm not messing up with the code flow and readability. Anyways, pushing the suggested changes now. :)

Thanks for making the code more readable. For now, I've tested your changes and it's working fine. I'll add more when needed.

paras-lehana commented:
> How does the 0 precision rounding work for viewing prices in USD where the difference between a price of $0.99 and $0.01 are substantial? Would this round everything up to $1 or down to $0 as that could be very misleading?
> 
> Actually the whole price is now cast to an int so all rounding would be lost no matter the precision?

You're right. Actually, I though decimal places in BTC and ETH waste space in the top bar so I had changed the precision to 0 from 2. It seems the precision should be decided by user itself so I have added a support for defining precision in `tokens.yaml`. The default precision is assumed 2 (as per your code). 

For example, my tokens.yaml looks like this: 

```yaml
tokens:
  Bitcoin,USD,0
  Ethereum,,2
  bgogo-token,SAT
```

The widget looks like this for this configuration:

![image](https://user-images.githubusercontent.com/10489912/56407830-07125c80-628f-11e9-8cb6-80c43baf1cbe.png)

I have pushed the required changes to the branch with final commit https://github.com/peter-featherstone/Ubuntu-Crypto-Price-Ticker/pull/1/commits/a9fe80c07670ee392d0c7de91020806ce3fd3d2c. Apologies for replying late this time. Do review and share more observations. I feel happy to contribute. :)

peter-featherstone commented:
This is looking excellent now! The only other improvement I can think of is instead of using comma separated lists for the yaml config to switch to a more traditional yaml config style such as:

```
tokens:
  Ethereum:
    currency: USD
    precision: 2
  Bitcoin:
     currency: GBP
  bgogo-token
```

This way the config uses explicit naming so is easier to build and if you leave out one of the config options you get the default value.

Sorry for all the suggestions, what you've done is great! If you don't want to make this change, that's fine I'm happy to merge as it is now.

paras-lehana commented:
Your suggestions are actually improving the code. I can happily follow your suggestions forever! ;)

Sure, I will need to read how to access intended properties. I have always used JSON and expecting this to be similar. As I'm out of the station, will only be able to push changes by Tuesday. :)
